cmd/kube-controller-manager/app/core.go

```diff
@@ -118,13 +120,32 @@ func startNodeIpamController(ctx ControllerContext) (http.Handler, bool, error)
 		}
 	}
 
+	nodeCIDRMaskSizes, err := processNodeCIDRMasks(ctx.ComponentConfig.NodeIPAMController.NodeCIDRMaskSize)
+	if err != nil {
+		return nil, false, err
+	}
+	// failure: length of node cidr masks greater than length of cluster cidrs
+	if len(nodeCIDRMaskSizes) > len(clusterCIDRs) {
+		return nil, false, fmt.Errorf("len of node cidr masks is %v > len of cluster cidrs", len(nodeCIDRMaskSizes))
+	}
+	// populate default mask size for each cluster cidr that doesn't have mask defined
+	if len(nodeCIDRMaskSizes) < len(clusterCIDRs) {
+		for _, clusterCIDR := range clusterCIDRs[len(nodeCIDRMaskSizes):] {
```

Comment:
So first CIDR mask is always for first cluster CIDR and use default for 2nd?

Reply:
Wondering if we should make it more explicit and require 2 arguments for `--node-cidr-mask-size` for dual-stack?

Reply:
most people are ok with the default sizes, specially for the ipv6 one. Forcing users to do all or none might be frowned upon.

Reply:
If no mask sizes are explicitly defined, instead of using `24` for all, we'll set the default sizes based on the cidr ip family. And yes, there is a one to one mapping between node cidr masks and cluster cidr.

Reply:
I still think this should be more explicit, for example this technically should error but I don't think it does?
```
--cluster-cidr=<ipv4-cidr>,<ipv6-cidr>
--node-cidr-masks=98
```

Reply:
but I would assume that works because this works:
```
--cluster-cidr=<ipv4-cidr>,<ipv6-cidr>
--node-cidr-masks=24
```

Reply:
I agree the semantics here are weird.  Parallel arrays are always clunky, but to do better would require some much more significant CLI surgery.  E.g. it might be cleaner or more obvious to allow:

```
--cluster-cidr=<v4> --node-cidr-mask=24 --cluster-cidr=<v6> --node-cidr-mask=104
```

Specifically this would be positional.  The implication being that:

```
--cluster-cidr=<v4>   \
--cluster-cidr=<v6>   \
--node-cidr-mask=104
```

...would use the default value for the first CIDR.  I am not sure this is better, though and here's why:

I expect that, before long, we allow multiple cluster CIDRs in the same family, to allow non-contiguous growth of a cluster.  I did some probing into this and there's actually only a handful of places that assume a single CIDR or even need to know about it at all.

Given that, we'll eventually have config like:

```
--cluster-cidr=<v4>,<v6>,<v4>,<v4>,<v4>
```

Now, if I have multiple cluster CIDRs of the same family - do I really want different node mask sizes for different CIDRs?  I don't think so.

So maybe what we really want is something more explicit, like:

```
--cluster-cidr=<v4>,<v6>   \
--node-cidr-mask-ipv4=25   \
--node-cidr-mask-ipv6=102
```

?

Reply:
I agree users would want to have one mask size per family. I think having distinct flags for each ip family is a good idea. Working out a few details before I make the change -

We would introduce the 2 new flags `--node-cidr-mask-size-ipv4` and `--node-cidr-mask-size-ipv6` in addition to the existing `--node-cidr-mask-size` flag.

1. If all 3 flags are provided -
```
--node-cidr-mask-size 24
--node-cidr-mask-size-ipv4 16
--node-cidr-mask-size-ipv6 98
```
Then v4/v6 mask sizes will be 16/98.

2. If `--node-cidr-mask-size` provided and either ipv4/ipv6 provided -
```
--node-cidr-mask-size 24
--node-cidr-mask-size-ipv4 16
```
then v4/v6 mask sizes would be 16/24 respectively.

```
--node-cidr-mask-size 24
--node-cidr-mask-size-ipv6 102
```
then v4/v6 mask sizes would be 24/102 respectively.

If `--node-cidr-mask-size` is provided and either of the other flags aren't provided, the `--node-cidr-mask-size` value will be used as default as that's what the user explicitly set.

3. If `--node-cidr-mask-size` is not set explicitly by user, then we would have internal defaults for ipv4 which is 24 and ipv6 - 98.

@thockin do you think this is reasonable?

Reply:
Changing the Cluster CIDR associated with `--node-cidr-mask-size` based on whether the other two flags `--node-cidr-mask-size-ipv4` and `--node-cidr-mask-size-ipv6` are set seems error prone to me. Wondering if this works:

* if single stack, only use `--node-cidr-mask-size`, error if either other two flags are set
* if dual-stack, only use `--node-cidr-mask-size-ipv4` and `--node-cidr-mask-size-ipv6` , error if `--node-cidr-mask-size` if used.

That way there's only these permutations of the flag that work:
```
--cluster-cidr=<ipv4>
--node-cidr-mask-size=24
```

```
--cluster-cidr=<ipv6>
--node-cidr-mask-size=98
```

```
--cluster-cidr=<ipv4,ipv6>
--node-cidr-mask-size-ipv4=24
--node-cidr-mask-size-ipv6=98
```

Reply:
I agree with Andrew's guidance here.  `--node-cidr-mask-size` should be mutually exclusive to the both of the new flags.

Now, should we deprecate the old flag and suggest people use the new ones, even in single-stack?  It's cleaner but more obnoxious.  We've generally stayed away from splitting flags into v4 and v6 variants (e.g. no `--cluster-cidr-ipv4` and `--cluster-cidr-ipv6`).

@khenidak

Reply:
If someone wants to use `--node-cidr-mask-size-ipv4` on a single stack deployment, I think the system should let them.

Reply:
I agree, we can keep it mutually exclusive. Also, like @thockin suggested, it's highly unlikely that a user would require different mask sizes for different CIDRs. Maybe in the future we can deprecate `--node-cdir-mask-size` and ask users to only use `--node-cidr-mask-size-ipv4` and `--node-cidr-mask-size-ipv6` for single stack/dual stack?

Reply:
updated the PR with the changes!